### PR TITLE
[WIP] Refactor `add` method and other

### PR DESCRIPTION
**Pull request by cristianoc72**
Starting from the discussion on #35 here are the most important refactors:

- [x]  Remove `phootwork\lang\parts\AddAllPart` containing the `addAll` method.
- [x]  Remove `phootwork\lang\parts\RemovePart::removeAll` method and refactor `phootwork\lang\parts\RemovePart::remove` method to accept one or more elements.
- [x] Refactor `phootwork\lang\parts\Add::add` method to accept one or more elements. Now the signature is `add(...$elements)`. Since we can't specify the index of the element anymore, we introduced the following method.
- [x] Introduce `phootwork\lang\parts\InsertPart::insert`method, which add one single element at the given index. The signature is `insert($element, $index)`. $index can be an int or a string, so it works on associative arrays, too.
- (REMOVED) See the discussion. ~~In case of associative arrays, give to `insert` method the possibility to define the index in a sort of "dot syntax". I.e. `insert('good item', 'content.items.goods')` which means `$array['content']['items']['goods'] = 'good item';~~
- [x] Refactor `phootwork\lang\parts\IndexFindersPart` methods to correctly work with associative arrays, too.
- [x] Add some more tests to verify the `phootwork\lang` library behavior with associative arrays

**Note**: this refactor introduces a disadvantage compared to the `addAll` method, when we operate with associative arrays: we can't add more elements at a time, because `...` operator can't unpack associative arrays.

### CODE REVIEW

**Comment by gossi** [comment]
I love the `insert()` and how it enables all of the other changes - I'm super in favor. What do you mean by the "dot syntax"/associative arrays to which construct (`ArrayObject`, `ArrayList` ?) it shall apply?

**Comment by cristianoc72** [comment]
After some reflection about the "dot syntax" I decided to suspend it. It's a much bigger refactor than I thought because it involves the whole library, introducing many complexities that I'm not sure they are profitable.
`ArrayObject` class and `collection` objects manipulate the "first level" of multidimensional arrays: if we introduce a way to manipulate also the deep keys, we have to refactor the finders, the index manipulation methods and so on. Is it profitable? Is it useful? Do we need it? Imho the answer is *no*.
Our library is already a gem as is, it works like a charm and we can add this functionality later, if we'll need it or someone else will ask for it.

**Comment by gossi** [comment]
I can confirm this. I'm using ember.js for frontend which also had this model. However, they moved away from it and it is sooo much better now with cleaner objects. Second is, that the dot-syntax is different in php. And concatenation here doesn't make that much sense, so it would be another bummer in that direction.

However, I see that if somebody needs this, they will provide proxy objects on top of it and if it really makes sense, they will perhaps publish it on packagist, too. For phootwork I'd totally agree with you, to keep it close to the language itself.

File: src/collection/AbstractList.php [hunk 1]
```diff
@@ -38,4 +46,63 @@ abstract class AbstractList extends AbstractCollection {
...
+	 * Returns the index of the given element or null if the element can't be found
+	 *
+	 * @param mixed $element
+	 *
+	 * @return int the index for the given element
+	 */
+	public function indexOf($element): ?int {
+		$index = $this->parentIndexOf($element);
+
+		return null === $index ? $index : (int) $index;
```
Comment (gossi) [inline src/collection/AbstractList.php]: Please no yoda style... hard to read it is.

File: src/collection/AbstractList.php [hunk 2]
```diff
@@ -38,4 +46,63 @@ abstract class AbstractList extends AbstractCollection {
...
+	 *     - find(callback)
+	 *
+	 * @param array $arguments
+	 *
+	 * @return int|null the index or null if it hasn't been found
+	 */
+	public function findLastIndex(...$arguments): ?int {
+		$lastIndex = $this->parentFindLastIndex(...$arguments);
+
+		return null === $lastIndex ? $lastIndex : (int) $lastIndex;
```
Comment (gossi) [inline src/collection/AbstractList.php]: Please no yoda style... hard to read it is.

File: src/collection/AbstractList.php [hunk 3]
```diff
@@ -38,4 +46,63 @@ abstract class AbstractList extends AbstractCollection {
...
+	 *     - find(callback)
+	 *
+	 * @param array $arguments
+	 *
+	 * @return int|null the index or null if it hasn't been found
+	 */
+	public function findIndex(...$arguments): ?int {
+		$index = $this->parentFindIndex(...$arguments);
+
+		return null === $index ? $index : (int) $index;
```
Comment (gossi) [inline src/collection/AbstractList.php]: Please no yoda style... hard to read it is.

File: src/collection/AbstractList.php [hunk 4]
```diff
@@ -23,7 +23,15 @@
  * @author Cristiano Cinotti
  */
 abstract class AbstractList extends AbstractCollection {
-	use EachPart, IndexFindersPart, ReducePart, RemovePart, ReversePart;
+	use EachPart;
+	use IndexFindersPart {
+		indexOf as parentIndexOf;
+		findLastIndex as parentFindLastIndex;
+		findIndex as parentFindIndex;
```
Comment (gossi) [inline src/collection/AbstractList.php]: Is there a common understanding of prefixing a mapping to a trait function with `parent*` that I don't know? If I read `$this->parent*` in the code later I would relate it to a method of the super class. If there is such a common methodology I'm fine and I learned it right now, anyway would suggest something like `trait*`?  What you think?

File: src/lang/parts/IndexFindersPart.php [hunk 5]
```diff
@@ -15,16 +15,16 @@ trait IndexFindersPart {
...
 	 *
-	 * @return int the index for the given element
+	 * @return int|string|null the index for the given element
 	 */
-	public function indexOf($element): ?int {
+	public function indexOf($element) {
 		$out = array_search($element, $this->array, true);
 
-		return false === $out ? null : (int) $out;
+		return false === $out ? null : $out;
```
Comment (gossi) [inline src/lang/parts/IndexFindersPart.php]: Please no yoda style... hard to read it is.

**Review by gossi — CHANGES_REQUESTED** [review]
Overall, pretty solid. I have a comment left about renaming methods from trait and please no yoda style (it's even funnier if you read these comments with original yoda voice in your head 😂

File: src/collection/AbstractList.php (hunk 4 above)
Comment (cristianoc72) [reply src/collection/AbstractList.php]: Whops, I forgot to find a better name. `trait*` is good for me.

File: src/collection/AbstractList.php (hunk 1 above)
Comment (cristianoc72) [reply src/collection/AbstractList.php]: :rofl: fixed.

File: src/collection/AbstractList.php (hunk 2 above)
Comment (cristianoc72) [reply src/collection/AbstractList.php]: fixed

File: src/collection/AbstractList.php (hunk 3 above)
Comment (cristianoc72) [reply src/collection/AbstractList.php]: fixed

File: src/collection/AbstractList.php (hunk 4 above)
Comment (cristianoc72) [reply src/collection/AbstractList.php]: fixed

File: src/lang/parts/IndexFindersPart.php (hunk 5 above)
Comment (cristianoc72) [reply src/lang/parts/IndexFindersPart.php]: fixed

**Review by gossi — APPROVED** [review]

**Comment by cristianoc72** [comment]
Time to draft 2.0 release!

**Pull request merged**